### PR TITLE
CTPPS update Run3 Direct Simulation to use old LHCInfo after LHCInfo update

### PR DESCRIPTION
**Pull request by Glitchmin**
#### PR description:

A followup PR to https://github.com/cms-sw/cmssw/pull/42515, this PR updates `Direct Simulation` to be able to use both old `LHCInfo` and new `LHCInfoPer*` records for possible future updates. 
Moreover it introduces a new Process Modifier `Run3_CTPPS_directSim` to be used when `Direct Simulation` is ran in Run3. This is needed because this simulation uses old `LHCInfo` no matter the run number. (In every other case modules ran in Run3 should be looking for new `LHCInfoPer*` records)

#### PR validation:

To validate please run Validation/CTPPS/test/simu/run_multiple

### CODE REVIEW

File: Configuration/Eras/python/Modifier_run3_directSim_cff.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import FWCore.ParameterSet.Config as cms
+
+run3_directSim =  cms.Modifier()
```
Comment (AndreaBellora) [inline Configuration/Eras/python/Modifier_run3_directSim_cff.py:3]: run3_directSim might not be a nice name. We should have the same behaviour also in Run 2 (or anyway make it clear that Run 2 works in the same way).

File: Validation/CTPPS/plugins/CTPPSHepMCDistributionPlotter.cc [hunk 2]
```diff
@@ -65,9 +76,23 @@ CTPPSHepMCDistributionPlotter::CTPPSHepMCDistributionPlotter(const edm::Paramete
...
+  desc.add<std::string>("outputFile", "")->setComment("output file");
+
+  descriptions.add("ctppsHepMCDistributionPlotter", desc);
+}
+
 void CTPPSHepMCDistributionPlotter::analyze(const edm::Event &iEvent, const edm::EventSetup &iSetup) {
   // get conditions
-  const auto &lhcInfo = iSetup.getData(lhcInfoESToken_);
+  LHCInfoCombined lhcInfoCombined(
+      iSetup, lhcInfoPerLSToken_, lhcInfoPerFillToken_, lhcInfoToken_, useNewLHCInfo_, iEvent.isRealData());
```
Comment (AndreaBellora) [inline Validation/CTPPS/plugins/CTPPSHepMCDistributionPlotter.cc:95]: This module is plotting HepMC stuff, so it's clearly running on MC. It won't even run on top of real data, so we can hard-code the usage of the old LHCInfo

File: CondTools/RunInfo/interface/LHCInfoCombined.h [hunk 3]
```diff
@@ -37,13 +38,16 @@ class LHCInfoCombined {
...
-      bool useNewLHCInfo);
+      bool useNewLHCInfo,
+      bool isRealData);
 
   void setFromLHCInfo(const LHCInfo& lhcInfo);
   void setFromPerLS(const LHCInfoPerLS& infoPerLS);
   void setFromPerFill(const LHCInfoPerFill& infoPerFill);
 
-  float crossingAngle();
+  static inline bool useNewLHCInfoForDirectSimu = false;
```
Comment (AndreaBellora) [inline CondTools/RunInfo/interface/LHCInfoCombined.h:48]: Do I understand correctly, this is a temporary way to switch between one mechanism and the other, right?
If so, it should be removed before the final PR. Also, I believe this should be const 😄

File: Configuration/Eras/python/Modifier_run3_directSim_cff.py (hunk 1 above)
Comment (Glitchmin) [reply Configuration/Eras/python/Modifier_run3_directSim_cff.py:3]: Hmm, that might be a good suggestion, if I understand correctly I can just name it `run_direct_sim` and it will be setting `useNewLHCInfo` to `false`

File: Validation/CTPPS/plugins/CTPPSHepMCDistributionPlotter.cc (hunk 2 above)
Comment (Glitchmin) [reply Validation/CTPPS/plugins/CTPPSHepMCDistributionPlotter.cc:95]: Well it didn't cost me much time to change it and now it offers some extra stuff like `isCrossingAngleValid` etc and maybe - that's actually something I don't know - `Direct Simulation` will one day be using new `LHCInfoPer*` records?

File: CondTools/RunInfo/interface/LHCInfoCombined.h (hunk 3 above)
Comment (Glitchmin) [reply CondTools/RunInfo/interface/LHCInfoCombined.h:48]: It should be `const` 
If `isRealData()` will make it to the final PR this bool should be here unless we're 100% sure `Direct Simulation` will never be using new `LHCInfoPer*`

File: Configuration/Eras/python/Era_Run3_CTPPS_directSim_cff.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import FWCore.ParameterSet.Config as cms
+
+from Configuration.Eras.Era_Run3_cff import Run3
+from Configuration.Eras.Modifier_run3_directSim_cff import run3_directSim
+
+Run3_CTPPS_directSim = cms.ModifierChain(Run3,run3_directSim)
```
Comment (makortel) [inline Configuration/Eras/python/Era_Run3_CTPPS_directSim_cff.py:6]: If these Era+Modifier are not about data taking conditions, it would be better to move the modifier to `Configuration/ProcessModifiers/python`.

File: Configuration/Eras/python/Modifier_run3_directSim_cff.py (hunk 1 above)
Comment (perrotta) [reply Configuration/Eras/python/Modifier_run3_directSim_cff.py:3]: I would also add "pps" to the name of the modifier

Comment (makortel) [reply Configuration/Eras/python/Modifier_run3_directSim_cff.py:3]: > I would also add "pps" to the name of the modifier

Using `ctpps` name would be more consistent with the existing modifiers.

File: Validation/CTPPS/plugins/CTPPSHepMCDistributionPlotter.cc (hunk 2 above)
Comment (AndreaBellora) [reply Validation/CTPPS/plugins/CTPPSHepMCDistributionPlotter.cc:95]: I don't think so, because the simulation should be setup via manual configurations. It's hard to say if run-dependent simulation will ever be used for PPS, but I'd exclude it as an option for now

File: Configuration/Eras/python/Modifier_run3_directSim_cff.py (hunk 1 above)
Comment (AndreaBellora) [reply Configuration/Eras/python/Modifier_run3_directSim_cff.py:3]: Yup, I believe that something like `ctpps_directSim` could be more appropriate, as we don't really distinguish between Run 2 and Run 3 for what concerns the usage of LHCInfo in the PPS direct simulation

**Comment by smuzaffar** [comment]
Milestone for this pull request has been moved to CMSSW_14_0_X.Please open a backport if it should also go in to CMSSW_13_3_X.

File: Configuration/Eras/python/Modifier_run3_directSim_cff.py (hunk 1 above)
Comment (Glitchmin) [reply Configuration/Eras/python/Modifier_run3_directSim_cff.py]: Changed in 861740283a6fa83d253f35682792539d48aa6c26

File: Validation/CTPPS/python/ctppsHepMCDistributionPlotter_cff.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+from Validation.CTPPS.CTPPSHepMCDistributionPlotter_cfi import *
+
+
+from Configuration.Eras.Modifier_run3_common_cff import run3_common
+run3_common.toModify(CTPPSHepMCDistributionPlotter, useNewLHCInfo = True)
```
[the remark this answers is not in the document]
Comment (perrotta) [reply Validation/CTPPS/python/ctppsHepMCDistributionPlotter_cff.py:5]: > @makortel Using `run3_common` was decided on a [AlCaDB meeting](https://indico.cern.ch/event/1308035/) with Tamas Vami and Francesco Brivio, where we explicitly discussed which modifier to use. `run3_common` is also used in a already closed PR #42515

@Glitchmin @tvami @francescobrivio could you please remind me which was the rationale of using `run3_common` instead of `ctpps_2022`? `run3_common` is built including `ctpps_2022` nonetheless, therefore the final effect should be the same. Using `run3_common` could be preferred in case a possible `ctpps_2024` (for example) will fully replace `ctpps_2022` (as it was the case for `ctpps_2022` that fully replaced previous `ctpps_2018` for run3): of course, in that case for future runN eras one would need to customize that ctpps_2024 explicitly, while by customizing `run3_common` instead this setting will be propagated till all future runs, including Phase2 ones.

[the remark this answers is not in the document]
Comment (tvami) [reply Validation/CTPPS/python/ctppsHepMCDistributionPlotter_cff.py:5]: @perrotta the point was that  `LHCInfoPer*` records are used in Run-3, and we don't plan to go back to Run-2/Run-1 with them (this is usually not the case when a new record is introduced). So putting that older PR under `run3_common` was very natural. Not sure what the (far) future plans (for a rereco) are for 900 GeV runs in 2021, but the `LHCInfoPer*` records should be populated there too, so using `ctpps_2022` might be misleading there. Otherwise probably `ctpps_2022` is good too.

[the remark this answers is not in the document]
Comment (perrotta) [reply Validation/CTPPS/python/ctppsHepMCDistributionPlotter_cff.py:5]: @tvami `ctpps_2022` is actually used to define the whole Run3 era, see [Era_Run3_cff.py](https://github.com/cms-sw/cmssw/blob/a35710c1d7e8d6d42cfe235bf92bdd5d57f573b7/Configuration/Eras/python/Era_Run3_cff.py): even in case of a possible rerun of 2021 data, that modifier will be picked by that era, in spite of having "2022" in its name. I agree that it would have been better to name `ctpps_2022` something like `run3_ctpps` since the beginning (as for the other subdetectors and subpackages): but here we are.

Therefore, for the modifications themselves using `ctpps_2022` or `run3_common` is completely equivalent. I tend to agree with @makortel that the "`run3_common` should not be used for subdetector-specific customizations". Beside that, I have not a strong opinion on that, and I wouldn't consider it a showstopper, unless @cms-sw/core-l2 people keep suggesting to enforce is.

**Comment by Glitchmin** [comment]
Since most of the doubts on topic of `Direct Simulation` and `useNewLHCInfo` have been cleared I'm closing this draft PR and moving the discussion to the official https://github.com/cms-sw/cmssw/pull/43397

**Comment by perrotta** [comment]
Cross posting from https://github.com/cms-sw/cmssw/pull/43397#issuecomment-1827601254: 
@Glitchmin is there any reasono not to modify this PR from draft to a full PR (so that all comments can remain saved in it) instead of opening a new PR and loose all history?

**Comment by Glitchmin** [comment]
> Cross posting from [#43397 (comment)](https://github.com/cms-sw/cmssw/pull/43397#issuecomment-1827601254): @Glitchmin is there any reasono not to modify this PR from draft to a full PR (so that all comments can remain saved in it) instead of opening a new PR and loose all history?

As discussed in https://github.com/cms-sw/cmssw/pull/43397 reopening this PR and turning it from draft to official

**Comment by perrotta** [comment]
please test

File: Validation/CTPPS/python/ctppsHepMCDistributionPlotter_cff.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (makortel) [reply Validation/CTPPS/python/ctppsHepMCDistributionPlotter_cff.py:5]: The design intent, that has also been mostly followed (but not everywhere, and not enforced), was that we create a Modifier for these "subdetector-specific" customizations, and then construct the Eras (technically ModifierChains) from these Modifiers kind of like building a house from lego bricks.

I realize now that the idea seems to be to change the default for Run3 and beyond to something, and then add another modifier to be able to switch the behavior back (to Run2 and earlier). If the customizations that are currently tied to `run3_common` would be done with another modifier, say `run3_ctpps_sim` (just for the example, please invent a better name), the `Run3_CTPPS_directSim` Era could be defined as
```py
Run3_CTPPS_directSim = Run3.copyAndExclude([run3_ctpps_sim])
```
To me this approach would be cleaner than customizing things back and forth.

File: CalibPPS/ESProducers/python/ctppsBeamParametersFromLHCInfoESSource_cff.py [hunk 6]
```diff
@@ -0,0 +1,7 @@
+from CalibPPS.ESProducers.ctppsBeamParametersFromLHCInfoESSource_cfi import *
+
+from Configuration.Eras.Modifier_run3_common_cff import run3_common
+run3_common.toModify(ctppsBeamParametersFromLHCInfoESSource, useNewLHCInfo = True)
```
Comment (makortel) [inline CalibPPS/ESProducers/python/ctppsBeamParametersFromLHCInfoESSource_cff.py:4]: Python objects should be modified only in the file they are defined in (otherwise various kind of weird behavior can occur in some cases). Simplest change would be

```py
from CalibPPS.ESProducers.ctppsBeamParametersFromLHCInfoESSource_cfi import ctppsBeamParametersFromLHCInfoESSource as _ctppsBeamParametersFromLHCInfoESSource
ctppsBeamParametersFromLHCInfoESSource = _ctppsBeamParametersFromLHCInfoESSource.clone()
```
but then there would be a conflict if both the `ctppsBeamParametersFromLHCInfoESSource_cfi` and `ctppsBeamParametersFromLHCInfoESSource_cff` were loaded into the same `cms.Process`. 

The `ctppsBeamParametersFromLHCInfoESSource_cfi` is generated by `fillDescriptions()`, right? Then an established pattern would be to generate the `cfi` under e.g. the name `ctppsBeamParametersFromLHCInfoESSourceDefault_cfi`, and do here
```py
from CalibPPS.ESProducers.ctppsBeamParametersFromLHCInfoESSourceDefault_cfi import ctppsBeamParametersFromLHCInfoESSourceDefault as _ctppsBeamParametersFromLHCInfoESSourceDefault
ctppsBeamParametersFromLHCInfoESSource = _ctppsBeamParametersFromLHCInfoESSourceDefault.clone()
```
Then this file could be named as `..._cfi.py`, and all the downstream could keep importing the `cfi` file.

Comment (makortel) [reply CalibPPS/ESProducers/python/ctppsBeamParametersFromLHCInfoESSource_cff.py:4]: > [#42515 (comment)](https://github.com/cms-sw/cmssw/pull/42515#discussion_r1297170431)
> `_cff` files that import `*` from `cfi` and use `run3_common.toModify()` have already been used in a merged PR #42515 and seem to cause no issue (https://github.com/cms-sw/cmssw/pull/42515/files#diff-cba86682cb8fea67d97d34b83f8883127a6c0f4f7b6d7454bccb7b4e29b43169)

This pattern is against our policy for the python fragment files (see 6.16 in https://cms-sw.github.io/cms_coding_rules.html#6--packaging-rules-1, also 6.13 and 6.14 are related). It may work today, but is brittle, and may easily break in the future.

Comment (perrotta) [reply CalibPPS/ESProducers/python/ctppsBeamParametersFromLHCInfoESSource_cff.py:4]: > I'm sorry but that comment seems to be in contrary to @perrotta's suggestion on how to handle those those modifiers - [#42515 (comment)](https://github.com/cms-sw/cmssw/pull/42515#discussion_r1297170431)

In doubt, always trust what @makortel says :-)

However, the two comments seem rather similar to me, besides the names given to the configuration files:
- I suggested to modify centrally a file named ctppsInterpolatedOpticalFunctionsESSource_cff.py that derives from a fillDescription generated ctppsInterpolatedOpticalFunctionsESSource_cfi.py
- Matti suggests to  modify centrally a file named ctppsInterpolatedOpticalFunctionsESSource_cfi.py that derives from a fillDescription generated ctppsBeamParametersFromLHCInfoESSourceDefault_cfi.py

Matti's suggested naming conventions are probably safer, in the sense that if implemented there is no risk that someone imports the unmodified fillDescription generated config thinking it is the already modified one instead.

Comment (makortel) [reply CalibPPS/ESProducers/python/ctppsBeamParametersFromLHCInfoESSource_cff.py:4]: > Matti's suggested naming conventions are probably safer, in the sense that if implemented there is no risk that someone imports the unmodified fillDescription generated config thinking it is the already modified one instead.

Right. The important part is that the file where the Modifiers are applied `clone()`s the python object. The `clone()` forms the definition of the python object, and then (by policy) the file is allowed to customize the object.

The pattern of `fillDescriptions()`-generated `ctppsInterpolatedOpticalFunctionsESSource_cfi.py`, from where `ctppsInterpolatedOpticalFunctionsESSource` is imported from and cloned in `ctppsInterpolatedOpticalFunctionsESSource_cff.py` would be ok for the policy. If someone would import/load the `cfi` in addition to the `cff` being used elsewhere, that would at least lead to framework raising an exception already during the python processing (of two modules attempting to use the same label).

With `fillDescriptions()` generating `<name>Default_cfi`, from where `<name>` is imported and `clone()`d in `<name>_cfi`, the probability for someone importing the `<name>Default_cfi` should be smaller. If both the `<name>Default_cfi` and `<name>_cfi` would be loaded into the same `cms.Process`, there would be no label collision, and the further behavior depends on the modules in question (for ES modules framework should throw an exception about two ES modules trying to produce the same things).

Comment (Glitchmin) [reply CalibPPS/ESProducers/python/ctppsBeamParametersFromLHCInfoESSource_cff.py]: I hope commit 4b795e5c13780fe4450986dbb9732f7d20b1826c  addresses this issue and makes the code a bit safer to use :)
Please take a look and let me know what do you think - if the code looks fine I'll merge the 2 commits

Comment (Glitchmin) [reply CalibPPS/ESProducers/python/ctppsBeamParametersFromLHCInfoESSource_cff.py]: Btw I had to create 2 `fillDescription` methods for my needs and did so using only my intuition - any advice would be welcome

**Comment by jfernan2** [comment]
please test

File: Validation/CTPPS/python/ctppsHepMCDistributionPlotter_cfi.py [hunk 7]
```diff
@@ -0,0 +1,8 @@
+from Validation.CTPPS.CTPPSHepMCDistributionPlotterDefault_cfi import CTPPSHepMCDistributionPlotterDefault as _CTPPSHepMCDistributionPlotterDefault
+CTPPSHepMCDistributionPlotter = _CTPPSHepMCDistributionPlotterDefault.clone()
+
+from Configuration.Eras.Modifier_run3_common_cff import run3_common
+run3_common.toModify(CTPPSHepMCDistributionPlotter, useNewLHCInfo = True)
+
+from Configuration.Eras.Modifier_ctpps_directSim_cff import ctpps_directSim
+ctpps_directSim.toModify(CTPPSHepMCDistributionPlotter, useNewLHCInfo = False)
```
Comment (perrotta) [inline Validation/CTPPS/python/ctppsHepMCDistributionPlotter_cfi.py:8]: All these could be written everywhere in a more compact way as
```suggestion
from Configuration.Eras.Modifier_run3_common_cff import run3_common
from Configuration.Eras.Modifier_ctpps_directSim_cff import ctpps_directSim
(run3_common & ~ctpps_directSim).toModify(CTPPSHepMCDistributionPlotter, useNewLHCInfo = True)
```

but I don't know if it suits your needs: take it as a possible suggestion, not as a request.

File: RecoPPS/ProtonReconstruction/python/ctppsProtons_cff.py [hunk 8]
```diff
@@ -7,11 +7,14 @@
 from CalibPPS.ESProducers.ctppsOpticalFunctions_cff import *
 
 # import and adjust proton-reconstructions settings
-from RecoPPS.ProtonReconstruction.ctppsProtons_cfi import *
-
+from RecoPPS.ProtonReconstruction.ctppsProtonsDefault_cfi import ctppsProtonsDefault as _ctppsProtonsDefault
```
[the remark this answers is not in the document]
Comment (Glitchmin) [reply RecoPPS/ProtonReconstruction/python/ctppsProtons_cff.py:10]: Without `from CalibPPS.ESProducers.ctppsOpticalFunctions_cff import *` in `ctppsProtons_cfi`  `TestCTPPSDirectProtonSimulation` is crashing with an error 
```No "CTPPSInterpolatedOpticsRcd" record found in the EventSetup.```

[the remark this answers is not in the document]
Comment (Glitchmin) [reply RecoPPS/ProtonReconstruction/python/ctppsProtons_cff.py:10]: However `from CalibPPS.ESProducers.ctppsAlignment_cff import *` seems to do nothing - as far as I can test it

File: Validation/CTPPS/python/ctppsHepMCDistributionPlotter_cfi.py (hunk 7 above)
Comment (Glitchmin) [reply Validation/CTPPS/python/ctppsHepMCDistributionPlotter_cfi.py:8]: Well I do admit it is more compact but definitely reduces code clarity - at least in my opinion. Using `~` symbol makes the code harder to understand and misreading just one symbol completety inverts the meaning of the code :P

File: RecoPPS/ProtonReconstruction/python/ctppsProtons_cff.py [hunk 9]
```diff
@@ -7,11 +7,14 @@
...
+ctppsProtons = _ctppsProtonsDefault.clone()
 
 ctppsProtons.pixelDiscardBXShiftedTracks = True
 ctppsProtons.default_time = -999.
```
[the remark this answers is not in the document]
Comment (Glitchmin) [reply RecoPPS/ProtonReconstruction/python/ctppsProtons_cff.py]: Done, in a file [`RecoPPS/ProtonReconstruction/python/ctppsProtons_cfi.py`](https://github.com/cms-sw/cmssw/blob/996e2c205386562c17759335a893f9a4da74e7be/RecoPPS/ProtonReconstruction/python/ctppsProtons_cfi.py)

File: RecoPPS/ProtonReconstruction/python/ctppsProtons_cff.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Glitchmin) [reply RecoPPS/ProtonReconstruction/python/ctppsProtons_cff.py]: Done, in a file [`RecoPPS/ProtonReconstruction/python/ctppsProtons_cfi.py`](https://github.com/cms-sw/cmssw/blob/996e2c205386562c17759335a893f9a4da74e7be/RecoPPS/ProtonReconstruction/python/ctppsProtons_cfi.py)

File: RecoPPS/ProtonReconstruction/python/ctppsProtons_cfi.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import FWCore.ParameterSet.Config as cms
+
+# import default optics settings
+from CalibPPS.ESProducers.ctppsOpticalFunctions_cff import *
```
Comment (perrotta) [inline RecoPPS/ProtonReconstruction/python/ctppsProtons_cfi.py:4]: Please make a RecoPPS/ProtonReconstruction/python/ctppsProtons_cff.py that includes both:
-  CalibPPS.ESProducers.ctppsOpticalFunctions_cff
- this  RecoPPS.ProtonReconstruction.ctppsProtons_cfi

as I suggested in https://github.com/cms-sw/cmssw/pull/42890#discussion_r1423730194 (before the "HOWEVER", as you verified that you cannot remove ctppsOpticalFunctions from it)

**Comment by perrotta** [comment]
@Glitchmin I appreciate that you squash the commits every time, However, as such it is hard to follow what had been modified and where. Please just squash the commit when the changes are all settled down: thank you!

**Comment by Glitchmin** [comment]
> @Glitchmin I appreciate that you squash the commits every time, However, as such it is hard to follow what had been modified and where. Please just squash the commit when the changes are all settled down: thank you!

I see, no problem!
The only modification this time is removal of `RecoPPS/ProtonReconstruction/python/ctppsProtons_cff.py` and addition of `RecoPPS/ProtonReconstruction/python/ctppsProtons_cfi.py` with applied changes (and replacing every `ctppsProtons_cff` occurence with `ctppsProtons_cfi`)

**Comment by civanch** [comment]
please test

**Comment by perrotta** [comment]
@Glitchmin [this](https://github.com/cms-sw/cmssw/pull/42890#discussion_r1424336914) still needs to be addressed

File: RecoPPS/ProtonReconstruction/python/ctppsProtons_cfi.py (hunk 10 above)
Comment (Glitchmin) [reply RecoPPS/ProtonReconstruction/python/ctppsProtons_cfi.py]: I hope commit 95349b14ebd2b18765811fc25ddda8b6000620f6 answers this comment - I just didn't know where to put `import FWCore.ParameterSet.Config as cms` but as far as I understood it should be in `_cff`

**Comment by perrotta** [comment]
> I hope commit [95349b1](https://github.com/cms-sw/cmssw/commit/95349b14ebd2b18765811fc25ddda8b6000620f6) answers this comment -

Thank you @Glitchmin!

>  I just didn't know where to put `import FWCore.ParameterSet.Config as cms` but as far as I understood it should be in `_cff`

Honestly, as far as I can see it is not needed in both of them: the `cms` shortcut is not used in any of these files, only in the ones from which you clone or import. So technically (1) you can remove it from both them. If we want to be picky, we can do it before the last squash.

And if we want to be even pickier, (2) you could also add a carriage return at the end of those files which are marked with a red sign in https://github.com/cms-sw/cmssw/pull/42890/files

Whatever you do (1), (2), both, or none of them + squash is fine with me (my preferred option: (1)+(2)+squash, of course): after that for what I am concerned this PR is good to go and can be signed by AlCa/DB

**Comment by Glitchmin** [comment]
> > I hope commit [95349b1](https://github.com/cms-sw/cmssw/commit/95349b14ebd2b18765811fc25ddda8b6000620f6) answers this comment -
> 
> Thank you @Glitchmin!
> 
> > I just didn't know where to put `import FWCore.ParameterSet.Config as cms` but as far as I understood it should be in `_cff`
> 
> Honestly, as far as I can see it is not needed in both of them: the `cms` shortcut is not used in any of these files, only in the ones from which you clone or import. So technically (1) you can remove it from both them. If we want to be picky, we can do it before the last squash.
> 
> And if we want to be even pickier, (2) you could also add a carriage return at the end of those files which are marked with a red sign in https://github.com/cms-sw/cmssw/pull/42890/files
> 
> Whatever you do (1), (2), both, or none of them + squash is fine with me (my preferred option: (1)+(2)+squash, of course): after that for what I am concerned this PR is good to go and can be signed by AlCa/DB

I don't know how it works but when I remove `import FWCore.ParameterSet.Config as cms` `TestCTPPSDirectProtonSimulation` crashes with an error
`ModuleNotFoundError: No module named 'IOMC.EventVertexGenerators.beamDivergenceVtxGenerator_cfi'`. I find it confusing since one should not affect the other, but it happens every time.

**Comment by Glitchmin** [comment]
> > I hope commit [95349b1](https://github.com/cms-sw/cmssw/commit/95349b14ebd2b18765811fc25ddda8b6000620f6) answers this comment -
> 
> Thank you @Glitchmin!
> 
> > I just didn't know where to put `import FWCore.ParameterSet.Config as cms` but as far as I understood it should be in `_cff`
> 
> Honestly, as far as I can see it is not needed in both of them: the `cms` shortcut is not used in any of these files, only in the ones from which you clone or import. So technically (1) you can remove it from both them. If we want to be picky, we can do it before the last squash.
> 
> And if we want to be even pickier, (2) you could also add a carriage return at the end of those files which are marked with a red sign in https://github.com/cms-sw/cmssw/pull/42890/files
> 
> Whatever you do (1), (2), both, or none of them + squash is fine with me (my preferred option: (1)+(2)+squash, of course): after that for what I am concerned this PR is good to go and can be signed by AlCa/DB

I pushed a squashed commit that doesn't remove `import FWCore.ParameterSet.Config as cms` because of the reasons in a previous message and adds a new line at the end of each file marked with the red sign

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
> I don't know how it works but when I remove `import FWCore.ParameterSet.Config as cms` `TestCTPPSDirectProtonSimulation` crashes with an error
> `ModuleNotFoundError: No module named 'IOMC.EventVertexGenerators.beamDivergenceVtxGenerator_cfi'`. I find it confusing since one should not affect the other, but it happens every time.

That's weird, but out of my comprehension
Maybe I'll try to find out why when I'll have some spare time, just to satisfy my curiosity,
As of now, just keep things as they are and avoid troubles.

**Comment by perrotta** [comment]
+1
- All issues raised during the review have been addressed

**Comment by makortel** [comment]
> > I don't know how it works but when I remove `import FWCore.ParameterSet.Config as cms` `TestCTPPSDirectProtonSimulation` crashes with an error
> > `ModuleNotFoundError: No module named 'IOMC.EventVertexGenerators.beamDivergenceVtxGenerator_cfi'`. I find it confusing since one should not affect the other, but it happens every time.
> 
> That's weird, but out of my comprehension Maybe I'll try to find out why when I'll have some spare time, just to satisfy my curiosity,

I agree this behavior sounds weird, and it would be great if you @perrotta would have some time to look into it.

File: CalibPPS/ESProducers/python/ctppsBeamParametersFromLHCInfoESSource_cff.py (hunk 6 above)
Comment (makortel) [reply CalibPPS/ESProducers/python/ctppsBeamParametersFromLHCInfoESSource_cff.py]: > Btw I had to create 2 `fillDescription` methods for my needs and did so using only my intuition - any advice would be welcome

Do you mean having one in `CTPPSProtonReconstructionEfficiencyEstimatorMC` and another in `CTPPSProtonReconstructionSimulationValidator`?

**Comment by jfernan2** [comment]
+1

**Comment by perrotta** [comment]
> > > I don't know how it works but when I remove `import FWCore.ParameterSet.Config as cms` `TestCTPPSDirectProtonSimulation` crashes with an error
> > > `ModuleNotFoundError: No module named 'IOMC.EventVertexGenerators.beamDivergenceVtxGenerator_cfi'`. I find it confusing since one should not affect the other, but it happens every time.
> > 
> > 
> > That's weird, but out of my comprehension Maybe I'll try to find out why when I'll have some spare time, just to satisfy my curiosity,
> 
> I agree this behavior sounds weird, and it would be great if you @perrotta would have some time to look into it.

Indeed, I am not able to reproduce.

I run
`scram b runtests`
from within Validation/CTPPS both with this PR as it is, and by removing 
`import FWCore.ParameterSet.Config as cms`
from RecoPPS/ProtonReconstruction/python/ctppsProtons_cff.py

In both cases the tests succeed.

@Glitchmin could you please check, and give us a recipe to reproduce that error if it is still there?

File: CalibPPS/ESProducers/plugins/CTPPSBeamParametersFromLHCInfoESSource.cc [hunk 11]
```diff
@@ -31,16 +31,23 @@ class CTPPSBeamParametersFromLHCInfoESSource : public edm::ESProducer {
...
 //----------------------------------------------------------------------------------------------------
 
 CTPPSBeamParametersFromLHCInfoESSource::CTPPSBeamParametersFromLHCInfoESSource(const edm::ParameterSet& iConfig)
-    : lhcInfoToken_(
-          setWhatProduced(this).consumes(edm::ESInputTag("", iConfig.getParameter<std::string>("lhcInfoLabel")))) {
+    : useNewLHCInfo_(iConfig.getParameter<bool>("useNewLHCInfo")) {
+  auto cc = setWhatProduced(this);
+  lhcInfoToken_ = cc.consumes(edm::ESInputTag("", iConfig.getParameter<std::string>("lhcInfoLabel")));
+  lhcInfoPerLSToken_ = cc.consumes(edm::ESInputTag("", iConfig.getParameter<std::string>("lhcInfoPerLSLabel")));
+  lhcInfoPerFillToken_ = cc.consumes(edm::ESInputTag("", iConfig.getParameter<std::string>("lhcInfoPerFillLabel")));
```
[the remark this answers is not in the document]
Comment (Glitchmin) [reply CalibPPS/ESProducers/plugins/CTPPSBeamParametersFromLHCInfoESSource.cc:49]: This does not sound like a bad idea, however in every module I worked on `lhcInfoLabel` is a string instead of `ESInputTag` so I think there might be a reason for it. It's worth investigating but I would defer it to a future PR - it's an aesthetic update, needed in many files and will take a while to implement and test

**Comment by Glitchmin** [comment]
> > > > I don't know how it works but when I remove `import FWCore.ParameterSet.Config as cms` `TestCTPPSDirectProtonSimulation` crashes with an error
> > > > `ModuleNotFoundError: No module named 'IOMC.EventVertexGenerators.beamDivergenceVtxGenerator_cfi'`. I find it confusing since one should not affect the other, but it happens every time.
> > > 
> > > 
> > > That's weird, but out of my comprehension Maybe I'll try to find out why when I'll have some spare time, just to satisfy my curiosity,
> > 
> > 
> > I agree this behavior sounds weird, and it would be great if you @perrotta would have some time to look into it.
> 
> Indeed, I am not able to reproduce.
> 
> I run `scram b runtests` from within Validation/CTPPS both with this PR as it is, and by removing `import FWCore.ParameterSet.Config as cms` from RecoPPS/ProtonReconstruction/python/ctppsProtons_cff.py
> 
> In both cases the tests succeed.
> 
> @Glitchmin could you please check, and give us a recipe to reproduce that error if it is still there?

I'm running `Validation/CTPPS/test/simu/run_multiple` which is the file ran when `TestCTPPSDirectProtonSimulation` when is ran. After removing the discussed line it crashed, but now

File: CalibPPS/ESProducers/plugins/CTPPSBeamParametersFromLHCInfoESSource.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (makortel) [reply CalibPPS/ESProducers/plugins/CTPPSBeamParametersFromLHCInfoESSource.cc:49]: > It's worth investigating but I would defer it to a future PR

That would be be perfectly fine, thanks.

(in principle using `ESInputTag` directly in configuration opens the possibility to use the first "module label" field of `ESInputTag` for ensuring the ES data products come from the expected ES module, if such requirement would be wanted, but that feature is not really used in CMSSW application configuration)

**Comment by Glitchmin** [comment]
Is there that needs to be done about this PR? Looking forward for comments and/or signatures :)

**Comment by perrotta** [comment]
@cms-sw/dqm-l2 @cms-sw/simulation-l2 please either comment or sign this PR, so that we can try to have it merged for pre2

**Comment by tjavaid** [comment]
@Glitchmin, @perrotta , we do not observe the [bin-by-bin comparison](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-13-2300+a9ac7a/60281/dqm-histo-comparison-summary.html) histograms to review before we could sign. There is an open [issue](https://github.com/cms-sw/cmssw/issues/43590) being discussed which is hoped to be fixed soon.

**Comment by civanch** [comment]
+1

**Comment by perrotta** [comment]
please test

**Comment by antoniovilela** [comment]
@cms-sw/dqm-l2 @cms-sw/alca-l2 
Do things look ok with this PR? 
Thanks.

**Comment by tjavaid** [comment]
+1

- spurious differences observed from unstable WF `12434.7` (open issue #39803 )

**Comment by antoniovilela** [comment]
+1
